ingester/src/deferred_load.rs

```diff
@@ -0,0 +1,369 @@
+                State::Unresolved(task_waker) => {
+                    // Wake the running background task, ignoring any send error
+                    // as the background task may have concurrently woken up due
+                    // to the sleep timer and stopped listening on the waker
+                    // channel.
+                    let _ = task_waker.send(());
+
+                    // Replace the state with a notification for this thread
+                    // (and others that call get()) to wait on for the
+                    // concurrent fetch to complete.
+                    Arc::new(Notify::default())
+                }
+
+                // If the value is already being fetched, wait for the fetch to
+                // complete.
+                State::Loading(waker) => waker,
+
+                // This was checked above before take()-ing the state.
+                State::Resolved(_) => unreachable!(),
+            };
+
+            // Ensure any subsequent callers can subscribe to the completion
+            // event.
+            *state = Some(State::Loading(Arc::clone(&waker)));
+
+            // Whenever there is a waker for the caller, the background task
+            // MUST be running.
+            //
+            // This check happens before the state lock is released, ensuring
+            // the background task doesn't concurrently finish (it would be
+            // blocked waiting to update the state).
+            assert!(!self.handle.is_finished());
+
+            waker
+        };
+
+        // Wait for the background task to complete resolving the value.
+        waker.notified().await;
+
+        match self.value.lock().as_ref().unwrap() {
+            State::Unresolved(_) | State::Loading(_) => unreachable!(),
+            State::Resolved(v) => v.clone(),
+        }
+    }
+}
+
+impl<T> Drop for DeferredLoad<T> {
+    fn drop(&mut self) {
+        // Attempt to abort the background task, regardless of it having
+        // completed or not.
+        self.handle.abort()
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use std::sync::Arc;
+
+    use futures::{executor::block_on, future, pin_mut};
+    use test_helpers::timeout::FutureTimeout;
+
+    use super::*;
+
+    /// Tests that want to exercise the demand loading configure the
+    /// DeferredLoad delay to this value, encouraging the background load to
+    /// happen a long time in the future.
+    ///
+    /// Because the delay before the background task runs is selected from a
+    /// uniform distribution, it is certainly possible the background task runs
+    /// before the demand call, invalidating the test - it's just statistically
+    /// very unlikely. Over multiple test runs, the odds of the background task
+    /// running before the demand call is so low, that if you see it happen
+    /// multiple times, I'd suggest taking out life insurance and staying
+    /// indoors during lightening storms.
```

Comment:
1. 🤣 
2. Typo, "lightening" should be "lightning"
3. This is a bit dark, isn't it? How about:

```suggestion
    /// multiple times, I'd suggest entering a lottery and/or checking changes for bugs.
```